### DISCUSSION TITLE
Client side redirects

### DISCUSSION THREAD

**Discussion by pmelab** [wakujs/waku#929] [General]
## The problem

When working with large volumes of content, redirects are used to retain internal and external integrity when paths change. For example 20k news articles potentially contain links to `/about` which is changed to `/about-us`. Even if the content management system could automatically update all references, it would mean that a lot of pages would have to be rebuilt and caches would be cleared. Redirects are a simple and efficient way to handle this.

When the HTTP server redirects a path from `/about` to `/about-us`, a Waku `<Link>` component pointing to `/about` will not follow the redirect, but display the `404` error page instead. Directly accessing `/about` will work as expected.

This topic has also been discussed in PR #895. Some different approaches from there, along with their caveats.

1. Make Waku router probe for a redirect before rendering a 404 page and load the redirect target RSC instead.

- Additional requests
- More complexity in the router

2. Load all possible redirects into the router.

- Potentially large amount of data
- More complexity in the router
- Requires a `createRedirect` API that can be used in `entries.tsx`

3. Create RSC payloads for redirected paths at build time.

- Increased amount of files (e.g. Cloudflare pages have a hard limit)
- Requires a `createRedirect` API that can be used in `entries.tsx`

4. Redirect `/RSC/about.txt` to `/RSC/about-us.txt` on the host system

- Leaks internal path of RSC to outside system
- Doubles the amount of necessary redirects (e.g. max 2000 on Cloudflare pages)

## PR with breaking test case
https://github.com/dai-shi/waku/pull/928

**Comment by dai-shi** [comment]
Not totally related, but can be related: #927

**Reply by pmelab** [reply]
Two sides of the same coin. If server side redirects are there, the Link (and imperative "navigate") should probably support them.

**Comment by dai-shi** [comment]
One of my concerns is Waku's responsibility for redirects. Particularly, one big use case of Waku is static sites.

Option 1: Waku or waku/router supports redirects natively
Option 2: Make waku/router RSC path be a spec (not an internal implementation detail)

**Reply by dai-shi** [reply]
Another requirement is we would like to disable SSR optionally.

**Reply by pmelab** [reply]
> Option 1: Waku or waku/router supports redirects natively

Thats my option "1" ...

> Option 2: Make waku/router RSC path be a spec (not an internal implementation detail)

... and thats basically "4".

I would prefer Option 1, because it has no dependencies on the host system. But I tried to implement it in the PR, and i failed ☹️

**Reply by dai-shi** [reply]
With option 1, I meant something like `createRedirect`.

**Reply by pmelab** [reply]
Then we would also have to provide an option to "export" those redirects (e.g. a server config). To have one source of truth.

**Reply by dai-shi** [reply]
So, does that mean host will do the redirects? That's different from what I mean by "natively".

**Reply by pmelab** [reply]
Maybe we should define what "natively" means 😄 
I always assume a static site as a baseline, so there is no node process to redirect. Therefore it has to be the host.

Gatsby's "createRedirect" is also able to output redirect configuration files for various hosting environments. But I'm not sure if that should really be a part of the framework. "Batteries included" is a costly catchphrase.

**Reply by dai-shi** [reply]
alright, that's probably lack of my knowledge.
I don't think we should go that direction for now. So, my option 1 is dropped.

**Comment by dai-shi** [comment]
How do other framework deal with redirects with client navigation?

**Reply by pmelab** [reply]
I know that Gatsby falls back to a `window.location.href = ...`. It's not perfect, since it fill force a full reload, but so far it was good enough. Just loading the correct RSC would definitely be better.

Will do a little case study. I was looking for an excuse to play with Astro anyway.

**Reply by rmarscher** [reply]
NextJS seems to just use their router.push or router.replace for clients-side redirects and tell you to call a server action for server coordinated redirects (which I understand Waku's unstable_redirect() supports). https://nextjs.org/docs/app/api-reference/functions/redirect#client-component

**Reply by pmelab** [reply]
I quickly tested Astro and NextJS with static builds and [serve](https://www.npmjs.com/package/serve):

* Astro elegantly dodges the problem by not having client side navigation in the first place  🥷 🤷 
* NextJS does the same as Gatsby. A top level reload if loading the data-file fails.
* For Remix this does not apply, redirects are handled on loaders on the server.

... any other frameworks we are interested in?

**Reply by rmarscher** [reply]
Thanks for looking into it. We found that sending a refresh meta tag is another option when a server component needs to tell the client to redirect. https://github.com/dai-shi/waku/discussions/927#discussioncomment-10820854

**Reply by dai-shi** [reply]
Thanks for checking it out!
We may reconsider this in the future (like after v1), but for now (towards v1), I think this is basically out of the scope of Waku.

My plan is the following:
- Make RSC path a spec for both minimal api and waku/router. (I'll make a change from what it is now.)
- So, it's up to host system how it redirects.
  - It would be nice to add some docs. (help wanted)
  - In the future, maybe some deploy adapters can generate redirect config automatically. (help wanted)
- Optionally, consider a way to attach a custom user logic so that it allows client side redirects.

**Reply by pmelab** [reply]
> My plan is the following: ...

The plan sounds great, but it's unclear to me if its in- or out of scope 🫤 

> (I'll make a change from what it is now.)

Is that change pre- or post v1?

**Reply by pmelab** [reply]
> Make RSC path a spec for both minimal api and waku/router. (I'll make a change from what it is now.)

I added the current state of this to the tests of #928. I also started some documentation there that can be adjusted accordingly. Just to have it written down somewhere.

**Reply by dai-shi** [reply]
> it's unclear to me if its in- or out of scope

I felt the same. More precisely, it's in the scope for the concept, but out of the scope for our code.

> Is that change pre- or post v1?

I'll work on it after releasing v0.21.3. I hope to finalize it, but we may revisit it based on feedback, hopefully before v1.

**Reply by rmarscher** [reply]
> * Make RSC path a spec for both minimal api and waku/router. (I'll make a change from what it is now.)

I don't have an issue with `/RSC/*`, but it also might be possible to place .txt and .html side-by-side in the static output and then check the accept header in the server. https://github.com/facebook/react/blob/main/fixtures/flight-esm/server/global.js#L63

**Reply by dai-shi** [reply]
It doesn't feel very comfortable to rely only on the accept header. tbh, I'm not super happy with `/RSC/*` and putting side-by-side sounds interesting, but still having `/RSC/` would be easier to understand and implement and should be less error-prune.
